### PR TITLE
NMD-413 Split dtos to entities, models, viewmodels

### PR DESCRIPTION
**Pull request by glebtyanov**

### CODE REVIEW

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementService.BLL/Services/ReviewService.cs [hunk 1]
```diff
@@ -1,21 +1,25 @@
...
+		public ReviewService(IReviewRepository repository, IMapper mapper)
 		{
 			_repository = repository;
+			_mapper = mapper;
 		}
 
 		public async Task<List<Review>> GetAllAsync()
 		{
-			return await _repository.GetAllAsync();
+			return _mapper.Map<List<Review>>(await _repository.GetAllAsync());
```
Comment (Glipchik) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementService.BLL/Services/ReviewService.cs:22]: Is it possible to use more abstract collection (i.e. IEnumerable, IQuerable, IList)?

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementService.BLL/Services/ReviewService.cs [hunk 2]
```diff
@@ -1,21 +1,25 @@
...
+		private readonly IMapper _mapper;
 
-		public ReviewService(IReviewRepository repository)
+		public ReviewService(IReviewRepository repository, IMapper mapper)
 		{
 			_repository = repository;
+			_mapper = mapper;
 		}
 
 		public async Task<List<Review>> GetAllAsync()
```
Comment (Glipchik) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementService.BLL/Services/ReviewService.cs:20]: The same

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementService.BLL/Services/ScheduleService.cs [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
+		private readonly IMapper _mapper;
 
-		public ScheduleService(IScheduleRepository repository)
+		public ScheduleService(IScheduleRepository repository, IMapper mapper)
 		{
 			_repository = repository;
+			_mapper = mapper;
 		}
 
 		public async Task<List<Schedule>> GetAllAsync()
```
Comment (Glipchik) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementService.BLL/Services/ScheduleService.cs:20]: The same and also in the next examples

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/ReviewsController.cs [hunk 4]
```diff
@@ -9,43 +12,43 @@ namespace SecondMap.Services.StoreManagementService.API.Controllers
...
-		public async Task<IActionResult> AddAsync([FromBody] Review reviewToAdd)
+		public async Task<IActionResult> AddAsync([FromBody] ReviewViewModel reviewToAdd)
 		{
-			await _reviewService.AddReviewAsync(reviewToAdd);
+			await _reviewService.AddReviewAsync(_mapper.Map<Review>(reviewToAdd));
 
 			return Ok();
 		}
 
 		[HttpPut("{id}")]
```
Comment (Glipchik) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/ReviewsController.cs:45]: Move strings to constants

**Review by Glipchik — CHANGES_REQUESTED** [review]

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/ReviewsController.cs [hunk 5]
```diff
@@ -1,56 +1,60 @@
...
+using SecondMap.Services.StoreManagementService.BLL.Constants;
 using SecondMap.Services.StoreManagementService.BLL.Interfaces;
-using SecondMap.Services.StoreManagementService.DAL.Models;
+using SecondMap.Services.StoreManagementService.BLL.Models;
 
 namespace SecondMap.Services.StoreManagementService.API.Controllers
 {
-	[Route("api/[controller]")]
+	[Route(ApiEndpoints.API_CONTROLLER_ROUTE)]
 	[ApiController]
```
Comment (Genresal) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/ReviewsController.cs:12]: Not an issue, but not sure that is necessary, to store routes inside constants. If you think that way then let it be. (some kind of design)

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/UsersController.cs [hunk 6]
```diff
@@ -1,56 +1,59 @@
...
+	[Route(ApiEndpoints.API_CONTROLLER_ROUTE)]
 	[ApiController]
 	public class UsersController : ControllerBase
 	{
 		private readonly IUserService _userService;
-
-		public UsersController(IUserService userService)
+		private readonly IMapper _mapper;
+		public UsersController(IUserService userService, IMapper mapper)
 		{
```
Comment (Genresal) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/UsersController.cs:18]: losted space between rows

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/UsersController.cs [hunk 7]
```diff
@@ -1,56 +1,59 @@
...
 
 		[HttpPost]
-		public async Task<IActionResult> AddAsync([FromBody] User userToAdd)
+		public async Task<IActionResult> AddAsync([FromBody] UserViewModel userToAdd)
 		{
-			await _userService.AddUserAsync(userToAdd);
+			await _userService.AddUserAsync(_mapper.Map<User>(userToAdd));
 
 			return Ok();
 		}
```
Comment (Genresal) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/UsersController.cs:43]: possible there need to return created data too, like inside your the "update" endpoint. and inside other controllers too.

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/MappingProfiles/ViewModelsToModelsProfile.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+﻿using AutoMapper;
+using SecondMap.Services.StoreManagementService.API.Dto;
+using SecondMap.Services.StoreManagementService.API.ViewModels;
+using SecondMap.Services.StoreManagementService.BLL.Models;
+
+namespace SecondMap.Services.StoreManagementService.API.MappingProfiles
+{
```
Comment (Genresal) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/MappingProfiles/ViewModelsToModelsProfile.cs:7]: better store all automapper profiles at one place. You can move all your models to separate project

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementService.BLL/Extensions/ServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -15,6 +16,8 @@ public static void AddServices(this IServiceCollection services)
 			services.AddScoped<IScheduleService, ScheduleService>();
 			services.AddScoped<IStoreService, StoreService>();
 			services.AddScoped<IUserService, UserService>();
+
+			services.AddAutoMapper(typeof(ModelToEntityProfile));
 		}
```
Comment (Genresal) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementService.BLL/Extensions/ServiceCollectionExtensions.cs:21]: Why you have automapper injecting there and at Program.cs, better unite all stuff and inject as one huge piece.

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/ViewModels/StoreViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿namespace SecondMap.Services.StoreManagementService.API.ViewModels
+{
+	public class StoreViewModel
+	{
+		public string Name { get; set; } = null!;
+		public string Address { get; set; } = null!;
+		public decimal Price { get; set; }
```
Comment (Genresal) [inline SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/ViewModels/StoreViewModel.cs:7]: why not string.empty, or you can make it nullable

**Review by Genresal — CHANGES_REQUESTED** [review]

File: SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/ReviewsController.cs (hunk 5 above)
Comment (glebtyanov) [reply SecondMap.Services.StoreManagementService/SecondMap.Services.StoreManagementServices.API/Controllers/ReviewsController.cs:12]: @Glipchik recommended to not have any hardcoded strings in the project, maybe I took it bit too far, might change it later

**Review by Glipchik — APPROVED** [review]

**Review by Genresal — APPROVED** [review]

**Pull request merged**